File: sdks/python/tests/worker_fixture.py
import logging
import os
import subprocess
import time
from io import BytesIO
from threading import Thread
from typing import Callable, Generator

import psutil
import requests


def wait_for_worker_health() -> bool:
    worker_healthcheck_attempts = 0
    max_healthcheck_attempts = 25

    while True:
        if worker_healthcheck_attempts > max_healthcheck_attempts:
            raise Exception(
                f"Worker failed to start within {max_healthcheck_attempts} seconds"
            )

        try:
            requests.get("http://localhost:8001/health", timeout=5)

            return True
        except Exception:
            time.sleep(1)

        worker_healthcheck_attempts += 1


def log_output(pipe: BytesIO, log_func: Callable[[str], None]) -> None:
    for line in iter(pipe.readline, b""):
        print(line.decode().strip())


def hatchet_worker(
    command: list[str],
) -> Generator[subprocess.Popen[bytes], None, None]:
    logging.info(f"Starting background worker: {' '.join(command)}")

    proc = subprocess.Popen(
        command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, env=os.environ.copy()
    )

    # Check if the process is still running
    if proc.poll() is not None:
        raise Exception(f"Worker failed to start with return code {proc.returncode}")

    Thread(target=log_output, args=(proc.stdout, logging.info), daemon=True).start()
    Thread(target=log_output, args=(proc.stderr, logging.error), daemon=True).start()

    wait_for_worker_health()

    yield proc

    logging.info("Cleaning up background worker")

    parent = psutil.Process(proc.pid)
    children = parent.children(recursive=True)

    for child in children:
        child.terminate()

    parent.terminate()

    _, alive = psutil.wait_procs([parent] + children, timeout=5)

    for p in alive:
        logging.warning(f"Force killing process {p.pid}")
        p.kill()

File: sdks/ruby/examples/worker_fixture.rb
# frozen_string_literal: true

require "open3"
require "net/http"
require "logger"
require "timeout"

module HatchetWorkerFixture
  LOGGER = Logger.new($stdout)

  # Wait for the worker health check endpoint to respond
  #
  # @param port [Integer] Health check port
  # @param max_attempts [Integer] Maximum number of attempts
  # @return [Boolean] true if healthy
  # @raise [RuntimeError] if worker fails to start
  def self.wait_for_worker_health(port:, max_attempts: 25)
    attempts = 0

    loop do
      if attempts > max_attempts
        raise "Worker failed to start within #{max_attempts} seconds"
      end

      begin
        uri = URI("http://localhost:#{port}/health")
        response = Net::HTTP.get_response(uri)
        return true if response.code == "200"
      rescue StandardError
        # Worker not ready yet
      end

      sleep 1
      attempts += 1
    end
  end

  # Start a worker subprocess and wait for it to be healthy
  #
  # @param command [Array<String>] Command to run
  # @param healthcheck_port [Integer] Port for health checks
  # @yield [pid] Yields the process PID
  # @return [void]
  def self.with_worker(command, healthcheck_port: 8001)
    LOGGER.info("Starting background worker: #{command.join(' ')}")

    ENV["HATCHET_CLIENT_WORKER_HEALTHCHECK_PORT"] = healthcheck_port.to_s

    stdin, stdout, stderr, wait_thr = Open3.popen3(*command)
    pid = wait_thr.pid

    # Log output in background threads
    Thread.new do
      stdout.each_line { |line| puts line.chomp }
    rescue IOError
      # Stream closed
    end

    Thread.new do
      stderr.each_line { |line| $stderr.puts line.chomp }
    rescue IOError
      # Stream closed
    end

    wait_for_worker_health(port: healthcheck_port)

    yield pid
  ensure
    LOGGER.info("Cleaning up background worker (PID: #{pid})")

    if pid
      begin
        # Kill process group to get children too
        Process.kill("TERM", -Process.getpgid(pid))
      rescue Errno::ESRCH, Errno::EPERM
        # Process already gone
      end

      begin
        Timeout.timeout(5) { Process.wait(pid) }
      rescue Timeout::Error
        begin
          Process.kill("KILL", pid)
          Process.wait(pid)
        rescue Errno::ESRCH, Errno::ECHILD
          # Already gone
        end
      rescue Errno::ECHILD
        # Already reaped
      end
    end

    [stdin, stdout, stderr].each do |io|
      io&.close rescue nil
    end
  end
end
